crates/wasmtime/src/instance.rs

```diff
@@ -299,6 +299,23 @@ impl Instance {
     /// # Panics
     ///
     /// Panics if `store` does not own this instance.
+    ///
+    /// # Why does `get_export` take a mutable context?
+    ///
+    /// This method requires a mutable context because an instance's exports are
+    /// lazily populated. While `wasmtime` could use interior mutability to
+    /// paper over this, since getting an export is "logically" a non-mutating
+    /// operation, this would bring with it a new problem. `wasmtime` would have
+    /// to choose whether to use a mutex or `RefCell` for the interior
+    /// mutability; the former implies unnecessasry overhead for single-threaded
+    /// usage, while the latter would prohibit `Store` from implementing `Send`
+    /// and `Sync`, making multi-threaded usage impossible.
+    ///
+    /// Given these trade offs -- and because `wasmtime` will not stop lazily
+    /// populating exports, because this makes instantiation faster -- we
+    /// decided that the best option is to avoid interior mutability and expose
+    /// what's happening under the covers to users via requiring a mutable
+    /// context.
```

Comment:
Personally I find this style of documentation to be helpful for a pretty small subset of people. Most of this can be understood by Rust & Wasmtime experts, but that's not necessarily who's always reading the documentation. I would instead perhaps recommend a style where we explain why the mutable context is used here and some benefits of that, but I wouldn't go into many details of "if we did this other thing it would have these implications".

Basically I don't want to draw too too much attention to this design decision because it's relatively minor compared to the rest of Wasmtime, and while it's important to be aware of it it's not necessarily too useful I think to consider what alternate implementation strategies were rejected when thinking about this. (tbh I never even considered using a lock or a cell, the mutability here just seemed natural)

Reply:
> (tbh I never even considered using a lock or a cell, the mutability here just seemed natural)

I would argue that this is only because you understand that it is doing lazy population and mutation under the covers though. I think most people would expect that getting an export is just reading a field or doing a hash map lookup.

Otherwise good points.